### DISCUSSION TITLE
React Native 0.71 版本支持

### DISCUSSION THREAD

**Discussion by zhiqingchen** [NervJS/taro#13148] [RFC]
- 提案时间: (2023-02-02)
- 影响版本: (3.7)

## 概述

将默认的 RN 版本切换至 0.71，expo 版本切换至 [48](https://blog.expo.dev/expo-sdk-48-beta-is-now-available-7ee5713d0937)。
react navigation 升级到 [7.x](https://reactnavigation.org/docs/7.x/upgrading-from-6.x)
参考[Taro 3.6 版本支持 React Native 0.70](https://github.com/NervJS/taro/discussions/12468)

## 动机

[React Native 0.71 版本已于 2023-1-12 正式发布](https://reactnative.dev/blog/2023/01/12/version-071)。
Taro 将与 RN 社区保持同步，将默认的 RN 版本设置为 0.71。

## 使用案例

将 Taro 升级至最新版，将原生代码根据相关pr进行变更。

### 依赖变更

todo

### Taro 主仓库升级至 0.71

升级内容：

0. 将不兼容的RN依赖库进行升级，并验证其功能可用性。
6. 将 `@types/react-native` 及 `react-native` 进行版本更新。
7. 将 metro、@react-native-community/cli-plugin-metro 相关依赖与 0.71 使用的版本进行同步。
8. 将默认安装的RN版本进行更新 https://github.com/NervJS/taro/blob/next/packages/taro-cli/src/presets/platforms/rn.ts#L33

### Taro Native Shell 升级至 0.71

分支：https://github.com/NervJS/taro-native-shell/tree/0.71.0

参考 [upgrade-helper](https://react-native-community.github.io/upgrade-helper/?from=0.70.6&to=0.71.0)

### Taro Project Template 升级至 0.71

0. 更新代码同 Taro Native Shell 升级至 0.71。

### Taro Playground 升级至 0.71

分支：

https://github.com/wuba/taro-playground/tree/releases/v1.9.0

升级内容：

0. 版本号改为 v1.9.0。
0. 更新代码同 Taro Native Shell 升级至 0.71。

## 缺陷

- 0.71 版本发布时间较短，及时跟进 0.71 问题。
- 默认 js 引擎为 Hermes，debug工具为 flipper，如不需要可自行切换。

## 替代选择

N/A

## 适配策略

- Taro 主工程默认使用 0.71，保证旧版本 RN 仍正常使用新版本的 Taro。
- 有些包（如 metro）可能无法向下兼容，则需要用户使用 resolution 进行版本锁定，需要在版本更新时说明。

### ffi 安装报错
https://github.com/ffi/ffi/issues/814

### expo-image-picker 安卓下编译报错
https://github.com/expo/expo/pull/20908
需要将 targetSdkVersion、compileSdkVersion、buildToolsVersion 升级到33

### The project is using an incompatible version (AGP 7.3.1) of the Android Gradle plugin. Latest supported version is AGP 7.2.0

AS 建议升级到 2022.2.1

### 编译时报 expo-modules-core 相关错误

将 package.json 中的 expo-modules-core 删除

**Comment by zhiqingchen** [comment]
## Taro Native Shell 升级至 0.71 测试

对应分支

https://github.com/NervJS/taro-native-shell/tree/0.71.0

下载现有包进行验证

https://github.com/NervJS/taro-native-shell/releases/tag/v0.71.6-0

修改代码后

```
# 10 为自增数字
git tag v0.71.0-3
git push origin v0.71.0-3
```

查看打包是否正常进行

https://github.com/NervJS/taro-native-shell/actions

**Reply by iambool** [reply]
验证了 https://github.com/NervJS/taro-native-shell/releases/tag/v0.71.6-beta.0

**Comment by zhiqingchen** [comment]
## Taro 主仓库升级至 0.71

对应分支

https://github.com/NervJS/taro/tree/feat/react-native-71

包安装

3.6.5-canary.1

### 测试要点

无

**Comment by zhiqingchen** [comment]
## Taro Project Template 升级至 0.71 测试

对应分支

https://github.com/zhiqingchen/taro-project-templates/tree/v3.7

```
# 全局安装beta版本cli
yarn global add @tarojs/cli@canary

# 通过模板进行初始
taro init --template-source github:zhiqingchen/taro-project-templates#v3.7

# 初始化后安装依赖
yarn upgradePeerdeps
```

1. 修改 assemble_ios_debug.yml 配置
   1. APP_ID: ${{secrets.APP_ID}}
   2. TEAM_ID: ${{secrets.TEAM_ID}}
   3. PROVISIONING_PROFILE_SPECIFIER: ${{secrets.DEBUG_PROVISIONING_PROFILE_SPECIFIER}}
   4. CODE_SIGN_IDENTITY: Apple Development
2. 修改 assemble_ios_release.yml 配置
   1. APP_ID: ${{secrets.APP_ID}}
   2. TEAM_ID: ${{secrets.TEAM_ID}}
   3. PROVISIONING_PROFILE_SPECIFIER: ${{secrets.RELEASE_PROVISIONING_PROFILE_SPECIFIER}}
   4. CODE_SIGN_IDENTITY: Apple Distribution

```
# 将源设置为 shell 工程
git remote add origin git@github.com:NervJS/taro-native-shell.git

# 10 为自增数字
git tag v0.71.0-beta.0
git push origin v0.71.0-beta.0
```

查看打包是否正常进行

https://github.com/NervJS/taro-native-shell/actions

参考分支

https://github.com/NervJS/taro-native-shell/tree/test/template-0.71-0

**Comment by zhiqingchen** [comment]
React Native 0.70 使用 Taro 3.7
===
旧版本从 3.6 升级到 3.7 或使用 Taro 3.6 初始化后，将 taro 相关版本升级到 3.7。需要做兼容。

## 方案一
1. `yarn upgradePeerdeps`
2. 将`android/build.gradle` 文件中 compileSdkVersion、buildToolsVersion、targetSdkVersion 升级到 33

**Comment by zhiqingchen** [comment]
## Taro Playground 升级至 0.71 测试

对应分支

https://github.com/wuba/taro-playground/tree/releases/v1.9.0

包下载
https://github.com/wuba/taro-playground/releases/tag/v1.9.0-beta.13
